### PR TITLE
[core] Add `eslint-plugin-react-compiler` experimental version and rules

### PR DESCRIPTION
**Pull request by JCQuintas**
Added multiple ways to enable the eslint rule, globaly or by "package"

```
ENABLE_REACT_COMPILER_PLUGIN=true pnpm eslint

// Or use one of the package specific ones
```

Part of https://github.com/mui/material-ui/issues/42548

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -210,17 +241,62 @@ module.exports = {
...
+      true,
+      ENABLE_REACT_COMPILER_PLUGIN_DATA_GRID,
+    ),
+    ...buildPackageRestrictedImports(
+      '@mui/x-data-grid-premium',
+      'x-data-grid-premium',
+      true,
+      ENABLE_REACT_COMPILER_PLUGIN_DATA_GRID,
+    ),
     ...buildPackageRestrictedImports('@mui/x-data-grid-generator', 'x-data-grid-generator'),
```
Comment (aarongarciah) [inline .eslintrc.js:275]: Should `ENABLE_REACT_COMPILER_PLUGIN_DATA_GRID` also be passed here?

**Review by aarongarciah — APPROVED** [review]
Nice! Just a small comment.

Apart from that, beware of the problem I just stumbled upon: we can't use `eslint-disable` for rules that are not set up. See: https://github.com/mui/material-ui/pull/42566#issuecomment-2155133428

edit: maybe we should always include the plugin but turn the rules off. I'll see if that works.

edit 2: it didn't work 😢

**Comment by aarongarciah** [comment]
@JCQuintas I finally went with the approach of commenting out the `// eslint-disable-next-line react-compiler/react-compiler` comments until we enable `eslint-plugin-react-compiler` https://github.com/mui/material-ui/pull/42566#issue-2340732308

**Comment by JCQuintas** [comment]
> @JCQuintas I finally went with the approach of commenting out the `// eslint-disable-next-line react-compiler/react-compiler` comments until we enable `eslint-plugin-react-compiler` [mui/material-ui#42566 (comment)](https://github.com/mui/material-ui/pull/42566#issue-2340732308)

We didnt start acting on these issues yet, will keep that in mind. Thanks 👍

**Pull request merged**